### PR TITLE
Add `BlockHeader` and replace protocol params with id in wrapper

### LINKED ISSUE #1224: Remove ProtocolParams from BlockWrapper and precompute ID

### LINKED ISSUE #1227: Add a BlockHeader type
not a bad idea, I'm fine with that

_Originally posted by @Alexandcoats in https://github.com/iotaledger/iota-sdk/pull/1190#discussion_r1324487586_

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Adds the `BlockHeader` type and replaces the protocol parameters stored in the wrapper with the pre-calculated block id.

## Links to any relevant issues

- Closes #1224 
- Closes #1227

### CODE REVIEW

File: sdk/src/types/block/core/wrapper.rs [hunk 1]
```diff
@@ -34,9 +36,90 @@ pub struct BlockWrapper {
...
+            slot_commitment_id,
+            latest_finalized_slot,
+            issuer_id,
+        })
+    }
+}
+
+/// Represent the object that nodes gossip around the network.
+#[derive(Clone, Debug, Eq, PartialEq, Getters, CopyGetters)]
+
```
Comment (thibault-martinez) [inline sdk/src/types/block/core/wrapper.rs]:
```suggestion
```

File: sdk/src/types/block/core/wrapper.rs [hunk 2]
```diff
@@ -34,9 +36,90 @@ pub struct BlockWrapper {
     latest_finalized_slot: SlotIndex,
     /// The identifier of the account that issued this block.
     issuer_id: IssuerId,
+}
+
+impl BlockHeader {
+    /// The length of the block header.
+    pub const LENGTH: usize = size_of::<Self>();
+
+    pub(crate) fn hash(&self) -> [u8; 32] {
```
Comment (Thoralf-M) [inline sdk/src/types/block/core/wrapper.rs:46]: Why not have this pub?

Comment (DaughterOfMars) [reply sdk/src/types/block/core/wrapper.rs:46]: I dunno, doesn't feel like it's something that needs to be pub

File: sdk/src/types/block/error.rs [hunk 3]
```diff
@@ -248,7 +248,7 @@ impl fmt::Display for Error {
             Self::InvalidInputKind(k) => write!(f, "invalid input kind: {k}"),
             Self::InvalidInputCount(count) => write!(f, "invalid input count: {count}"),
             Self::InvalidInputOutputIndex(index) => write!(f, "invalid input or output index: {index}"),
-            Self::InvalidBlockLength(length) => write!(f, "invalid block length {length}"),
+            Self::InvalidBlockWrapperLength(length) => write!(f, "invalid block length {length}"),
```
Comment (thibault-martinez) [inline sdk/src/types/block/error.rs]:
```suggestion
            Self::InvalidBlockWrapperLength(length) => write!(f, "invalid block wrapper length {length}"),
```

File: sdk/src/types/block/core/wrapper.rs [hunk 4]
```diff
@@ -34,9 +36,89 @@ pub struct BlockWrapper {
...
+        })
+    }
+}
+
+/// Represent the object that nodes gossip around the network.
+#[derive(Clone, Debug, Eq, PartialEq, Getters, CopyGetters)]
+pub struct BlockWrapper {
+    /// The identifier of the block.
+    #[getset(get_copy = "pub")]
+    id: BlockId,
```
Comment (thibault-martinez) [inline sdk/src/types/block/core/wrapper.rs]: Can you put it last please?

Comment (DaughterOfMars) [reply sdk/src/types/block/core/wrapper.rs]: I mean, sure, but why? Feels like a first kind of field

Comment (thibault-martinez) [reply sdk/src/types/block/core/wrapper.rs]: Just feels right in terms of dependency, you need the other fields to compute it so I think last is more appropriate. Also adding at the bottom of the TIP just feels nicer.

Comment (DaughterOfMars) [reply sdk/src/types/block/core/wrapper.rs]: k done

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/types/block/core/wrapper.rs [hunk 5]
```diff
@@ -34,133 +36,173 @@ pub struct BlockWrapper {
...
+/// Represent the object that nodes gossip around the network.
+#[derive(Clone, Debug, Eq, PartialEq, Getters, CopyGetters)]
+pub struct BlockWrapper {
+    #[getset(skip)]
+    header: BlockHeader,
     /// The inner block.
+    #[getset(get = "pub")]
     block: Block,
     /// The block signature, used to validate issuance capabilities.
+    #[getset(get_copy = "pub")]
```
Comment (thibault-martinez) [inline sdk/src/types/block/core/wrapper.rs:119]: How does that work? Signature is not Copy ?

Comment (DaughterOfMars) [reply sdk/src/types/block/core/wrapper.rs:119]: It works because Signature IS Copy 😆

Comment (thibault-martinez) [reply sdk/src/types/block/core/wrapper.rs:119]: Arf, I was looking on develop 🤦🏻

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by abdulmth — APPROVED** [review]

**Pull request merged**